clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/IngestionThrottler.java

```diff
@@ -0,0 +1,151 @@
+ * mode when the bootstrapping is done.
+ */
+public class IngestionThrottler implements Closeable {
+  private final Logger LOGGER = LogManager.getLogger(IngestionThrottler.class);
+
+  private final static int CURRENT_VERSION_BOOTSTRAPPING_DEFAULT_CHECK_INTERVAL = 30;
+  private final static TimeUnit CURRENT_VERSION_BOOTSTRAPPING_DEFAULT_CHECK_TIMEUNIT = TimeUnit.SECONDS;
+
+  private final ScheduledExecutorService eventThrottlerUpdateService;
+
+  private volatile EventThrottler finalRecordThrottler;
+  private volatile EventThrottler finalBandwidthThrottler;
+  private boolean isUsingSpeedupThrottler = false;
+
+  public IngestionThrottler(
+      boolean isDaVinciClient,
+      VeniceServerConfig serverConfig,
+      Supplier<Map<String, StoreIngestionTask>> ongoingIngestionTaskMapSupplier) {
+    this(
+        isDaVinciClient,
+        serverConfig,
+        ongoingIngestionTaskMapSupplier,
+        CURRENT_VERSION_BOOTSTRAPPING_DEFAULT_CHECK_INTERVAL,
+        CURRENT_VERSION_BOOTSTRAPPING_DEFAULT_CHECK_TIMEUNIT);
+  }
+
+  public IngestionThrottler(
+      boolean isDaVinciClient,
+      VeniceServerConfig serverConfig,
+      Supplier<Map<String, StoreIngestionTask>> ongoingIngestionTaskMapSupplier,
+      int checkInterval,
+      TimeUnit checkTimeUnit) {
+
+    EventThrottler regularRecordThrottler = new EventThrottler(
+        serverConfig.getKafkaFetchQuotaRecordPerSecond(),
+        serverConfig.getKafkaFetchQuotaTimeWindow(),
+        "kafka_consumption_records_count",
+        false,
+        EventThrottler.BLOCK_STRATEGY);
+    EventThrottler regularBandwidthThrottler = new EventThrottler(
+        serverConfig.getKafkaFetchQuotaBytesPerSecond(),
+        serverConfig.getKafkaFetchQuotaTimeWindow(),
+        "kafka_consumption_bandwidth",
+        false,
+        EventThrottler.BLOCK_STRATEGY);
+
+    if (isDaVinciClient && serverConfig.isDaVinciCurrentVersionBootstrappingSpeedupEnabled()) {
+      EventThrottler speedupRecordThrottler = new EventThrottler(
+          serverConfig.getDaVinciCurrentVersionBootstrappingQuotaRecordsPerSecond(),
+          serverConfig.getKafkaFetchQuotaTimeWindow(),
+          "current_version_speedup_kafka_consumption_records_count",
+          false,
+          EventThrottler.BLOCK_STRATEGY);
+      EventThrottler speedupBandwidthThrottler = new EventThrottler(
+          serverConfig.getDaVinciCurrentVersionBootstrappingQuotaBytesPerSecond(),
+          serverConfig.getKafkaFetchQuotaTimeWindow(),
+          "current_version_speedup_kafka_consumption_bandwidth",
+          false,
+          EventThrottler.BLOCK_STRATEGY);
+      this.eventThrottlerUpdateService =
+          Executors.newSingleThreadScheduledExecutor(new DaemonThreadFactory("Ingestion_Event_Throttler_update"));
+      this.eventThrottlerUpdateService.scheduleAtFixedRate(() -> {
+        Map<String, StoreIngestionTask> ongoingStoreIngestionTaskMap = ongoingIngestionTaskMapSupplier.get();
```

Comment:
Should we exit here if we see isUsingSpeedupThrottler is set to false? Once all current version are bootstrapped we don't need to do this check anymore. I don't know if we can cancel the scheduling inside the schedule logic, but at least we can skip all these logics.

Reply:
Or is it valid because we HAVE TO periodically check, since there might be edge case that this check runs first before bootstrap() or active ingestion comes in and if the order does not play well we will not trigger the speed up throttler...

Reply:
The logic was designed on purpose in this way to handle the runtime current version, which shouldn't be very common.
Another case is that the customer might bootstrap stores one by one, so we don't want to miss any of them.
The check is quite cheap, so keeping it shouldn't introduce much overhead.

Reply:
OK, I think that makes sense to me